### PR TITLE
add sharding for gpt3 

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

New features
### PR changes

Models
### Description

add sharding parallel for gpt-3

### CODE REVIEW

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 1]
```diff
@@ -30,26 +30,20 @@
 import lr
 from paddle.distributed import fleet
 from paddle.distributed.fleet.meta_parallel import get_rng_state_tracker
+from paddle.distributed.fleet.meta_optimizers.dygraph_optimizer.dygraph_sharding_optimizer import DygraphShardingOptimizer
```
Comment (ZeyuChen) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: It's too long for this import ,is it ok to update the init file to shorter the import path?

**Review by ZeyuChen — COMMENTED** [review]
Please shorter the import path of `DygraphShardingOptimizer` since it's not so friendly for developer.

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 2]
```diff
@@ -30,26 +30,20 @@
...
-def set_hyrbid_parallel_seed(basic_seed, dp_rank, mp_rank, pp_rank):
+def set_hyrbid_parallel_seed(basic_seed, idx):
     assert args.device != "cpu"
 
-    random.seed(basic_seed + dp_rank)
-    np.random.seed(basic_seed + dp_rank)
-    paddle.seed(basic_seed + dp_rank)
-
-    # local_seed/ global_seed is used to control dropout in ModelParallel
-    local_seed = basic_seed + 123 + mp_rank * 10 + pp_rank * 1000
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:48]: @ForFishes 这里的seed删除没有关系吗？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 3]
```diff
@@ -30,26 +30,20 @@
...
 
 MODEL_CLASSES = {
     "gpt": (GPTForPretraining, GPTTokenizer),
     "gpt-cn": (GPTForPretraining, GPTChineseTokenizer),
 }
 
 
-def set_hyrbid_parallel_seed(basic_seed, dp_rank, mp_rank, pp_rank):
+def set_hyrbid_parallel_seed(basic_seed, idx):
     assert args.device != "cpu"
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:42]: 对，这里需要给一个 warning 吧， @ForFishes 后面可以加一下

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 4]
```diff
@@ -174,7 +184,7 @@ def do_train(args):
 
     clip = None
     if args.grad_clip > 0:
-        clip = paddle.nn.ClipGradByGlobalNorm(clip_norm=args.grad_clip)
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: ？为什么改掉`ClipGradByGlobalNorm`

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 5]
```diff
@@ -227,8 +253,8 @@ def do_train(args):
                 args,
                 data_file,
                 local_rank=local_rank,
-                data_world_size=args.dp_degree,
-                data_world_rank=dp_rank,
+                data_world_size=worker_num,
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 这里不能这么改，要考虑MP

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 6]
```diff
@@ -309,9 +335,9 @@ def do_train(args):
                                  args.eval_iters, log_writer, global_step,
                                  epoch, "valid")
 
-                # only dp_rank = 0 save model
+                # only worker_index = 0 save model
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 要考虑MP

**Review by ZHUI — CHANGES_REQUESTED** [review]

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 4 above)
Comment (JZ-LIANG) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 需要支持 ClipGradByGlobalNorm， 如果报错请查看 https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/distributed/fleet/meta_optimizers/dygraph_optimizer/hybrid_parallel_optimizer.py#L45  @zhaoyingli

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 5 above)
Comment (JZ-LIANG) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: data_world_size 只需要考虑读不同数据等rank？ dp 和 sharding group 读取不同的数据， mp 和 pp 读取相同的数据，所以data_world_size = dp_degree * sharding_degree

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 6 above)
Comment (JZ-LIANG) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 需要考虑mp 的什么？ 数据loader 还是 model save ?

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 5 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 我看岔了，`worker_index worker_num`你们重新定义了，sorry！

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 7]
```diff
@@ -108,10 +105,23 @@ def do_train(args):
     mp_rank = hcg.get_model_parallel_rank()
     pp_rank = hcg.get_stage_id()
     dp_rank = hcg.get_data_parallel_rank()
+    sharding_rank = hcg.get_sharding_parallel_rank()
+
+    sharding_size = hcg.get_sharding_parallel_world_size()
+    worker_index = dp_rank * sharding_size + sharding_rank
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: How about?
```
worker_index -> data_world_rank
worker_num -> data_world_size
```

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 2 above)
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:48]: seed 的东西 可否都移入到这里 @zhaoyuchen2018

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 8]
```diff
@@ -309,9 +335,9 @@ def do_train(args):
                                  args.eval_iters, log_writer, global_step,
                                  epoch, "valid")
 
-                # only dp_rank = 0 save model
+                # only worker_index = 0 save model
                 if (global_step % args.save_steps == 0 or
-                        global_step >= args.max_steps) and dp_rank == 0:
+                        global_step >= args.max_steps) and worker_index == 0:
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 静态图是专门写了 save_persistable 来save，这里的sharding，需要其他专门的save支持吗？

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 1 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: done.

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 4 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已支持ClipGradByGlobalNorm。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 7 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: done.

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 8 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: done.

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 9]
```diff
@@ -92,14 +86,17 @@ def do_train(args):
...
+        "pp_degree": args.pp_degree,
+        "sharding_degree": args.sharding_degree
     }
 
     strategy.pipeline_configs = {
         "accumulate_steps": args.local_batch_size // args.micro_batch_size,
         "micro_batch_size": args.micro_batch_size
     }
 
+    strategy.tensor_parallel_configs = {"tensor_init_seed": 123, }
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: what's `tensor_init_seed` for ?

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 10]
```diff
@@ -322,24 +348,25 @@ def do_train(args):
                     logger.info("Save model to %s" % output_dir)
 
                     if args.pp_degree > 1:
-                        model_to_save.save_state_dict(output_dir)
-                        if mp_rank * pp_rank == 1:
+                        if mp_rank == 0 and sharding_rank == 0 and pp_rank == 0:
```
Comment (ZHUI) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:352]: 所以目前的策略是，`dp_rank == 0` 下所有参数都保存。

现阶段，可以不加load 参数，但辛苦确认一下，load checkpoint会不会存在问题？

另外记一个合并参数的 TODO 吧

**Review by ZHUI — DISMISSED** [review]
LGTM

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 3 above)
Comment (ForFishes) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:42]: ok， will fix in next pr.

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 11]
```diff
@@ -108,10 +105,23 @@ def do_train(args):
...
+    sharding_rank = hcg.get_sharding_parallel_rank()
+
+    sharding_size = hcg.get_sharding_parallel_world_size()
+    data_world_rank = dp_rank * sharding_size + sharding_rank
+    data_world_size = args.dp_degree * args.sharding_degree
     local_rank = int(os.getenv("PADDLE_RANK_IN_NODE", 0))
 
     # seed control in hybrid parallel
-    set_hyrbid_parallel_seed(args.seed, dp_rank, mp_rank, pp_rank)
+    set_hyrbid_parallel_seed(args.seed, data_world_rank)
```
Comment (ForFishes) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py]: 所有seed都在这里设置吧

File: examples/language_model/gpt-3/dygraph/run_pretrain.py [hunk 12]
```diff
@@ -322,24 +348,25 @@ def do_train(args):
                     logger.info("Save model to %s" % output_dir)
 
                     if args.pp_degree > 1:
-                        model_to_save.save_state_dict(output_dir)
-                        if mp_rank * pp_rank == 1:
+                        if mp_rank == 0 and sharding_rank == 0 and pp_rank == 0:
                             tokenizer.save_pretrained(output_dir)
+                        model_to_save.save_state_dict(output_dir)
```
Comment (ForFishes) [inline examples/language_model/gpt-3/dygraph/run_pretrain.py:354]: 模型保存这里，没有详细的测试。需要记个TODO

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 9 above)
Comment (ZHUI) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: @ForFishes

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 12 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:354]: 已更新

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 11 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: 已更新

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 10 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py:352]: dp_rank==0下所有参数都保存这个逻辑没有问题，但是现在模型保存没有测试，所以可能会存在问题，比如是否存储了全量的参数。

File: examples/language_model/gpt-3/dygraph/run_pretrain.py (hunk 9 above)
Comment (zhaoyinglia) [reply examples/language_model/gpt-3/dygraph/run_pretrain.py]: set control in tensor parallel

**Review by ZHUI — APPROVED** [review]
LGTM
